From: Georgia Garcia <georgia.garcia@canonical.com>
Date: Mon, 16 Mar 2026 18:14:17 -0300
Subject: [PATCH 2/5] parser: convert conditionals operators to an enum

This will allow us to support different operators in the future

Signed-off-by: Georgia Garcia <georgia.garcia@canonical.com>
---
 parser/af_rule.cc     | 12 ++++++------
 parser/af_rule.h      |  2 +-
 parser/af_unix.cc     |  4 ++--
 parser/dbus.cc        |  6 +++---
 parser/io_uring.cc    |  6 +++---
 parser/mount.cc       | 24 ++++++++++++++++--------
 parser/mqueue.cc      |  6 +++---
 parser/network.cc     |  6 +++---
 parser/parser.h       | 10 ++++++++--
 parser/parser_misc.c  |  4 ++--
 parser/parser_regex.c |  2 +-
 parser/parser_yacc.y  |  8 ++++----
 parser/ptrace.cc      |  6 +++---
 parser/signal.cc      |  6 +++---
 parser/userns.cc      |  6 +++---
 15 files changed, 61 insertions(+), 47 deletions(-)

diff --git a/parser/af_rule.cc b/parser/af_rule.cc
index e384d95..eb42494 100644
--- a/parser/af_rule.cc
+++ b/parser/af_rule.cc
@@ -34,10 +34,10 @@
 
 /* need to move this table to stl */
 static struct supported_cond supported_conds[] = {
-	{ "type", true, false, false, local_cond },
-	{ "protocol", false, false, false, local_cond },
-	{ "label", true, false, false, peer_cond },
-	{ NULL, false, false, false, local_cond },	/* eol sentinel */
+	{ "type", true, cond_comp::EQ, false, local_cond },
+	{ "protocol", false, cond_comp::EQ, false, local_cond },
+	{ "label", true, cond_comp::EQ, false, peer_cond },
+	{ NULL, false, cond_comp::NONE, false, local_cond },	/* eol sentinel */
 };
 
 bool af_rule::cond_check(struct supported_cond *conds, struct cond_entry *ent,
@@ -53,8 +53,8 @@ bool af_rule::cond_check(struct supported_cond *conds, struct cond_entry *ent,
 			yyerror("%s rule: '%s' conditional is only valid in the peer expression\n", rname, ent->name);
 		if (peer && (i->side == local_cond))
 			yyerror("%s rule: '%s' conditional is not allowed in the peer expression\n", rname, ent->name);
-		if (!ent->eq && !i->in)
-			yyerror("%s rule: keyword 'in' is not allowed in '%s' socket conditional\n", rname, ent->name);
+		if (ent->comp != i->supported_conds)
+			yyerror("%s rule: operator used is not allowed in '%s' socket conditional\n", rname, ent->name);
 		if (list_len(ent->vals) > 1 && !i->multivalue)
 			yyerror("%s rule: conditional '%s' only supports a single value\n", rname, ent->name);
 		return true;
diff --git a/parser/af_rule.h b/parser/af_rule.h
index 4294490..de4c14f 100644
--- a/parser/af_rule.h
+++ b/parser/af_rule.h
@@ -32,7 +32,7 @@ enum cond_side { local_cond, peer_cond, either_cond };
 struct supported_cond {
 	const char *name;
 	bool supported;
-	bool in;
+	cond_comp supported_conds;
 	bool multivalue;
 	enum cond_side side ;
 };
diff --git a/parser/af_unix.cc b/parser/af_unix.cc
index bac6625..3a3de3b 100644
--- a/parser/af_unix.cc
+++ b/parser/af_unix.cc
@@ -42,8 +42,8 @@ int parse_unix_perms(const char *str_perms, perm32_t *perms, int fail)
 
 
 static struct supported_cond supported_conds[] = {
-	{ "addr", true, false, false, either_cond },
-	{ NULL, false, false, false, local_cond },	/* sentinel */
+	{ "addr", true, cond_comp::EQ, false, either_cond },
+	{ NULL, false, cond_comp::EQ, false, local_cond },	/* sentinel */
 };
 
 void unix_rule::move_conditionals(struct cond_entry *conds)
diff --git a/parser/dbus.cc b/parser/dbus.cc
index 6af32d4..dd5f3f9 100644
--- a/parser/dbus.cc
+++ b/parser/dbus.cc
@@ -40,9 +40,9 @@ void dbus_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in dbus rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of dbus rules\n");
 		if (list_len(cond_ent->vals) > 1)
 			yyerror("dbus conditional \"%s\" only supports a single value\n",
 				cond_ent->name);
diff --git a/parser/io_uring.cc b/parser/io_uring.cc
index 60b8c25..9f6777d 100644
--- a/parser/io_uring.cc
+++ b/parser/io_uring.cc
@@ -30,9 +30,9 @@ void io_uring_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in io_uring rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of io_uring rules\n");
 
 		if (list_len(cond_ent->vals) > 1)
 			yyerror("io_uring conditional \"%s\" only supports a single value\n",
diff --git a/parser/mount.cc b/parser/mount.cc
index 6a8b936..c584a4c 100644
--- a/parser/mount.cc
+++ b/parser/mount.cc
@@ -405,6 +405,9 @@ static struct value_list *extract_fstype(struct cond_entry **conds)
 	struct cond_entry *entry, *tmp, *prev = NULL;
 
 	list_for_each_safe(*conds, entry, tmp) {
+		if (entry->comp != cond_comp::EQ && entry->comp != cond_comp::IN)
+			yyerror("only \"=\" and 'in' allowed in conditions of mount rules\n");
+
 		if (strcmp(entry->name, "fstype") == 0 ||
 		    strcmp(entry->name, "vfstype") == 0) {
 			PDEBUG("  extracting fstype\n");
@@ -420,17 +423,20 @@ static struct value_list *extract_fstype(struct cond_entry **conds)
 	return list;
 }
 
-static struct cond_entry *extract_options(struct cond_entry **conds, int eq)
+static struct cond_entry *extract_options(struct cond_entry **conds, cond_comp comp)
 {
 	struct cond_entry *list = NULL, *entry, *tmp, *prev = NULL;
 
 	list_for_each_safe(*conds, entry, tmp) {
+		if (entry->comp != cond_comp::EQ && entry->comp != cond_comp::IN)
+			yyerror("only \"=\" and 'in' allowed in conditions of mount rules\n");
+
 		if ((strcmp(entry->name, "options") == 0 ||
 		     strcmp(entry->name, "option") == 0) &&
-		    entry->eq == eq) {
+		    entry->comp == comp) {
 			list_remove_at(*conds, prev, entry);
-			PDEBUG("  extracting %s %s\n", entry->name, entry->eq ? 
-"=" : "in");
+			PDEBUG("  extracting %s %s\n",
+			       entry->name, entry->comp == cond_comp::EQ ? "=" : " in ");
 			list_append(entry, list);
 			list = entry;
 		} else
@@ -445,7 +451,8 @@ static void perror_conds(const char *rule, struct cond_entry *conds)
 	struct cond_entry *entry;
 
 	list_for_each(conds, entry) {
-		PERROR(  "unsupported %s condition '%s%s(...)'\n", rule, entry->name, entry->eq ? "=" : " in ");
+		PERROR(  "unsupported %s condition '%s%s(...)'\n",
+			 rule, entry->name, entry->comp == cond_comp::EQ ? "=" : " in ");
 	}
 }
 
@@ -469,7 +476,8 @@ static void process_one_option(struct cond_entry *&opts, unsigned int &flags,
 	entry->vals = NULL;
 	/* fail if there are any unknown optional flags */
 	if (opts) {
-		PERROR("  unsupported multiple 'mount options %s(...)'\n", entry->eq ? "=" : " in ");
+		PERROR("  unsupported multiple 'mount options %s(...)'\n",
+		       entry->comp == cond_comp::EQ ? "=" : " in ");
 		exit(1);
 	}
 	free_cond_entry(entry);
@@ -493,7 +501,7 @@ mnt_rule::mnt_rule(struct cond_entry *src_conds, char *device_p,
 
 	if (src_conds) {
 		/* move options in () to local list */
-		struct cond_entry *opts_in = extract_options(&src_conds, 0);
+		struct cond_entry *opts_in = extract_options(&src_conds, cond_comp::IN);
 
 		if (opts_in) {
 			unsigned int tmpflags = 0, tmpinv_flags = 0;
@@ -512,7 +520,7 @@ mnt_rule::mnt_rule(struct cond_entry *src_conds, char *device_p,
 		}
 
 		/* move options=() to opts list */
-		struct cond_entry *opts_eq = extract_options(&src_conds, 1);
+		struct cond_entry *opts_eq = extract_options(&src_conds, cond_comp::EQ);
 		if (opts_eq) {
 			unsigned int tmpflags = 0, tmpinv_flags = 0;
 			struct cond_entry *entry;
diff --git a/parser/mqueue.cc b/parser/mqueue.cc
index 93eaa99..d44c0bb 100644
--- a/parser/mqueue.cc
+++ b/parser/mqueue.cc
@@ -63,9 +63,9 @@ void mqueue_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in mqueue rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of mqueue rules\n");
 
 		if (strcmp(cond_ent->name, "label") == 0) {
 			move_conditional_value("mqueue", &label, cond_ent);
diff --git a/parser/network.cc b/parser/network.cc
index 90557f4..b3faa76 100644
--- a/parser/network.cc
+++ b/parser/network.cc
@@ -404,9 +404,9 @@ void network_rule::move_conditionals(struct cond_entry *conds, ip_conds &ip_cond
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in network rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of network rules\n");
 		if (strcmp(cond_ent->name, "ip") == 0) {
 			move_conditional_value("network", &ip_cond.sip, cond_ent);
 			if (!parse_address(ip_cond))
diff --git a/parser/parser.h b/parser/parser.h
index ed444e3..386a08f 100644
--- a/parser/parser.h
+++ b/parser/parser.h
@@ -117,9 +117,15 @@ struct value_list {
 
 int cmp_value_list(value_list *lhs, value_list *rhs);
 
+enum class cond_comp {
+	NONE,
+	EQ,
+	IN,
+};
+
 struct cond_entry {
 	char *name;
-	int eq;			/* where equals was used in specifying list */
+	cond_comp comp; /* '=' or 'in' was used in specifying list */
 	struct value_list *vals;
 
 	struct cond_entry *next;
@@ -459,7 +465,7 @@ extern int is_blacklisted(const char *name, const char *path);
 extern struct value_list *new_value_list(char *value);
 extern void free_value_list(struct value_list *list);
 extern void print_value_list(struct value_list *list, std::ostream &os);
-extern struct cond_entry *new_cond_entry(char *name, int eq, struct value_list *list);
+extern struct cond_entry *new_cond_entry(char *name, cond_comp comp, struct value_list *list);
 extern void move_conditional_value(const char *rulename, char **dst_ptr,
 				   struct cond_entry *cond_ent);
 extern void free_cond_entry(struct cond_entry *ent);
diff --git a/parser/parser_misc.c b/parser/parser_misc.c
index 5da1fae..c6e112b 100644
--- a/parser/parser_misc.c
+++ b/parser/parser_misc.c
@@ -1216,13 +1216,13 @@ void move_conditional_value(const char *rulename, char **dst_ptr,
 	cond_ent->vals->value = NULL;
 }
 
-struct cond_entry *new_cond_entry(char *name, int eq, struct value_list *list)
+struct cond_entry *new_cond_entry(char *name, cond_comp comp, struct value_list *list)
 {
 	struct cond_entry *ent = (struct cond_entry *) calloc(1, sizeof(struct cond_entry));
 	if (ent) {
 		ent->name = name;
 		ent->vals = list;
-		ent->eq = eq;
+		ent->comp = comp;
 	}
 
 	return ent;
diff --git a/parser/parser_regex.c b/parser/parser_regex.c
index 248dfb6..b63c12a 100644
--- a/parser/parser_regex.c
+++ b/parser/parser_regex.c
@@ -450,7 +450,7 @@ const char *local_name(const char *name)
  */
 char *get_xattr_value(struct cond_entry *entry)
 {
-	if (!entry->eq)
+	if (entry->comp != cond_comp::EQ)
 		return NULL;
 	if (!entry->vals)
 		return NULL;
diff --git a/parser/parser_yacc.y b/parser/parser_yacc.y
index ab6fef5..95b05d1 100644
--- a/parser/parser_yacc.y
+++ b/parser/parser_yacc.y
@@ -1300,7 +1300,7 @@ network_rule: TOK_NETWORK opt_net_perm TOK_ID TOK_ID opt_conds opt_cond_list TOK
 cond: TOK_CONDID
 	{
 		struct cond_entry *ent;
-		ent = new_cond_entry($1, 0, NULL);
+		ent = new_cond_entry($1, cond_comp::NONE, NULL);
 		if (!ent)
 			yyerror(_("Memory allocation error."));
 		$$ = ent;
@@ -1312,7 +1312,7 @@ cond: TOK_CONDID TOK_EQUALS TOK_VALUE
 		struct value_list *value = new_value_list($3);
 		if (!value)
 			yyerror(_("Memory allocation error."));
-		ent = new_cond_entry($1, 1, value);
+		ent = new_cond_entry($1, cond_comp::EQ, value);
 		if (!ent) {
 			free_value_list(value);
 			yyerror(_("Memory allocation error."));
@@ -1322,7 +1322,7 @@ cond: TOK_CONDID TOK_EQUALS TOK_VALUE
 
 cond: TOK_CONDID TOK_EQUALS TOK_OPENPAREN valuelist TOK_CLOSEPAREN
 	{
-		struct cond_entry *ent = new_cond_entry($1, 1, $4);
+		struct cond_entry *ent = new_cond_entry($1, cond_comp::EQ, $4);
 
 		if (!ent)
 			yyerror(_("Memory allocation error."));
@@ -1332,7 +1332,7 @@ cond: TOK_CONDID TOK_EQUALS TOK_OPENPAREN valuelist TOK_CLOSEPAREN
 
 cond: TOK_CONDID TOK_IN TOK_OPENPAREN valuelist TOK_CLOSEPAREN
 	{
-		struct cond_entry *ent = new_cond_entry($1, 0, $4);
+		struct cond_entry *ent = new_cond_entry($1, cond_comp::IN, $4);
 
 		if (!ent)
 			yyerror(_("Memory allocation error."));
diff --git a/parser/ptrace.cc b/parser/ptrace.cc
index 2580f0a..29d78b7 100644
--- a/parser/ptrace.cc
+++ b/parser/ptrace.cc
@@ -34,9 +34,9 @@ void ptrace_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in ptrace rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of ptrace rules\n");
 
 		if (strcmp(cond_ent->name, "peer") == 0) {
 			move_conditional_value("ptrace", &peer_label, cond_ent);
diff --git a/parser/signal.cc b/parser/signal.cc
index 514a5a1..91312de 100644
--- a/parser/signal.cc
+++ b/parser/signal.cc
@@ -162,9 +162,9 @@ void signal_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in signal rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of signal rules\n");
 		if (strcmp(cond_ent->name, "set") == 0) {
 			extract_sigs(&cond_ent->vals);
 		} else if (strcmp(cond_ent->name, "peer") == 0) {
diff --git a/parser/userns.cc b/parser/userns.cc
index a2cd9e8..285eb3c 100644
--- a/parser/userns.cc
+++ b/parser/userns.cc
@@ -30,9 +30,9 @@ void userns_rule::move_conditionals(struct cond_entry *conds)
 	struct cond_entry *cond_ent;
 
 	list_for_each(conds, cond_ent) {
-		/* for now disallow keyword 'in' (list) */
-		if (!cond_ent->eq)
-			yyerror("keyword \"in\" is not allowed in userns rules\n");
+		/* for now allow only '=' */
+		if (cond_ent->comp != cond_comp::EQ)
+			yyerror("only \"=\" allowed in conditions of userns rules\n");
 
 		/* no valid conditionals atm */
 		yyerror("invalid userns rule conditional \"%s\"\n",
